src/Functions/formatDateTime.cpp

```diff
@@ -51,22 +69,82 @@ template <> struct ActionValueTypeMap<DataTypeDateTime>   { using ActionValueTyp
 // TODO(vnemkov): to add sub-second format instruction, make that DateTime64 and do some math in Action<T>.
 template <> struct ActionValueTypeMap<DataTypeDateTime64> { using ActionValueType = Int64; };
 
+/// Counts the number of literal characters in Joda format string until the next closing literal
+/// sequence single quote. Returns -1 if no literal single quote was found.
+/// In Joda format string(https://joda-time.sourceforge.net/apidocs/org/joda/time/format/DateTimeFormat.html)
+/// literal content must be quoted with single quote. and two single quote means literal with one single quote.
+/// For example:
+/// Format string: "'aaaa'", unescaped literal: "aaaa";
+/// Format string: "'aa''aa'", unescaped literal: "aa'aa";
+/// Format string: "'aaa''aa" is not valid because of missing of end single quote.
+Int64 numLiteralChars(const char * cur, const char * end)
+{
+    Int64 count = 0;
+    while (cur < end)
+    {
+        if (*cur == '\'')
+        {
+            if (cur + 1 < end && *(cur + 1) == '\'')
+            {
+                count += 2;
+                cur += 2;
+            }
+            else
+                return count;
+        }
+        else
+        {
+            ++count;
+            ++cur;
+            // No end literal single quote found
+            if (cur == end)
+                return -1;
+        }
+    }
+    return count == 0 ? -1 : count;
+}
+
+/// Cast content from integer to string, making sure digits number in result string is no less than total_digits by padding leading '0'.
+String padContent(UInt32 content, size_t total_digits)
+{
+    String str_content = std::to_string(content);
+    auto str_content_length = str_content.size();
+    if (str_content_length == 0)
```

Comment:
This special handling of empty content strings is subsumed by l. 115ff. You could remove l. 112/113 and it will continue to work.